services/storage/src/simcore_service_storage/simcore_s3_dsm.py

```diff
@@ -106,31 +108,85 @@ async def list_datasets(self, user_id: UserID) -> list[DatasetMetaData]:
         data: list[FileMetaData] = await self.list_files(
-            user_id, uuid_filter=dataset_id + "/"
+            user_id, expand_dirs, uuid_filter=ensure_ends_with(dataset_id, "/")
         )
         return data
 
+    async def _expand_directory(
+        self, fmd: FileMetaDataAtDB, max_items_to_include: NonNegativeInt
+    ) -> list[FileMetaData]:
+        """
+        Scans S3 backend and returns a list S3MetaData entries which get mapped
+        to FileMetaData entry.
+        """
+        files_in_folder: list[S3MetaData] = await get_s3_client(self.app).list_files(
+            self.simcore_bucket_name,
+            prefix=ensure_ends_with(fmd.file_id, "/"),
+            max_files_to_list=max_items_to_include,
+        )
+        result: list[FileMetaData] = [
+            convert_db_to_model(
+                FileMetaDataAtDB(
+                    location_id=fmd.location_id,
+                    location=fmd.location,
+                    bucket_name=fmd.bucket_name,
+                    object_name=x.file_id,
+                    user_id=fmd.user_id,
+                    # NOTE: to ensure users have a consistent experience the
+                    # `created_at` field is inherited from the last_modified
+                    # coming from S3. This way if a file is created 1 month after the
+                    # creation of the directory, the file's creation date
+                    # will not be 1 month in the passed.
+                    created_at=x.last_modified,
+                    file_id=x.file_id,
+                    file_size=parse_obj_as(ByteSize, x.size),
+                    last_modified=x.last_modified,
+                    entity_tag=x.e_tag,
+                    is_soft_link=False,
+                    is_directory=False,
+                )
+            )
+            for x in files_in_folder
+        ]
+        return result
+
     async def list_files(
-        self, user_id: UserID, uuid_filter: str = ""
+        self, user_id: UserID, expand_dirs: bool, uuid_filter: str = ""
     ) -> list[FileMetaData]:
+        """
+        expand_dirs `False`: returns the one metadata entry for each directory
+        expand_dirs `True`: returns all files in each directory (no directories will be included)
+
+        NOTE: expand_dirs will be replaced by pagination in the future
+        currently only {EXPAND_DIR_MAX_ITEM_COUNT} items will be returned
+        The endpoint produces similar results to what it did previously
+        """
+
         data: list[FileMetaData] = []
-        accesible_projects_ids = []
+        accessible_projects_ids = []
         async with self.engine.acquire() as conn, conn.begin():
-            accesible_projects_ids = await get_readable_project_ids(conn, user_id)
-            file_metadatas: list[
+            accessible_projects_ids = await get_readable_project_ids(conn, user_id)
+            file_meta_data: list[
```

Comment:
so you removed the s. ok since data is plural. but now these are all the folders. so maybe rename to folders_meta_data